### ISSUE TITLE
Kafka trigger don't listen but emits an error

### ISSUE THREAD

**Issue by bastianfbr** [n8n-io/n8n#10933]
### Bug Description

When enabling my Kafka Trigger workflow, it never listen my messages and i just have an error when i activate/test it : 
2024-09-23 14:24:49 {"level":"ERROR","timestamp":"2024-09-23T12:24:49.860Z","logger":"kafkajs","message":"[Connection] Response Heartbeat(key: 12, version: 3)","broker":"**_hidden for privacy_**","clientId":"n8n","error":"The group is rebalancing, so a rejoin is needed","correlationId":978,"size":10}

### To Reproduce

Create a Kafka Trigger workflow
Test or Activate it
Send a message

### Expected behavior

Kafka Trigger should be triggered by a message according to given topic.

### Operating System

Windows 11

### n8n Version

1.57.0

### Node.js Version

18

### Database

SQLite (default)

### Execution mode

main (default)

**Comment by Joffcom** [comment]
Hey @bastianfbr,

We have created an internal ticket to look into this which we will be tracking as "GHC-249"

**Comment by Joffcom** [comment]
Hey @bastianfbr,

It looks like Kafka is returning an error "The group is rebalancing, so a rejoin is needed", Are you using a Kafka cluster and what is the general health like as that message typically only appears when a new consumer is added or removed from a cluster, If that error keeps occuring it would suggest the cluster is unhappy with something as it likely shouldn't be always changing.

**Comment by bastianfbr** [comment]
I'm using a Kafka cluster. Everything's fine but trigger don't listen to kafka as it's never triggered...

**Comment by bastianfbr** [comment]
I understand what you're saying so i tested myself with a tiny kafkajs consumer/producer code and i also have 
> The group is rebalancing, so a rejoin is needed
but after, everything's fine, so it's just like a "warning" rather than an error...

**Comment by Joffcom** [comment]
Hey @bastianfbr, 

I wonder if our Kafka package treats it as an error because the cluster is saying it is an error 🤔 I would also not expect it to always happen which is strange. In n8n when the activation fails does it work if you try to activate it again after or just it always just get that error and fail?

**Comment by bastianfbr** [comment]
> Hey @bastianfbr,
> 
> I wonder if our Kafka package treats it as an error because the cluster is saying it is an error 🤔 I would also not expect it to always happen which is strange. In n8n when the activation fails does it work if you try to activate it again after or just it always just get that error and fail?

The activation doesn't fail
It just a log in my kafka and the trigger loops and doesn't listen, even if i send a message to the topic...
If i activate again and again, it still doesn't listen a single message.

**Comment by Joffcom** [comment]
Perfect hanks, I have updated the internal ticket so we will just need someone to pick it up and work on it.

**Comment by afunworm** [comment]
I'm facing the same problem and unable to get Kafka trigger to listen properly. Commenting to follow this thread.

I also have Bitnami Kafka in a docker, running along side n8n Docker on the same server.

If you have found any way to make this work, please let me know. Will hosting kafka on a separate server work?

**Comment by afunworm** [comment]
> @afunworm i’ll recommend few things
> 
> * do not use same consumer group more than once (diff consumer group for each trigger)
> * dont use test workflow in kafka trigger, just activate it and edit while it is active in production mode.
> * Deactivating the workflow sometimes doesnt deactivate unless we restart n8n container. So do not rely on deactivating trigger, just disconnect the trigger from pipeline when you need to stop.
> * in 1 out of 10000 cases, we get an error like No workflow data (even with webhooks trigger as well) (in queue mode usually)
> * Do not stop workflow on error in any node, just continue on error, so it will not stuck in loop
> 
> overall you will feel comfortable with time. Because not everything is ideal.
> 
> Good Luck.

Thank you SO MUCH. These are really sound advice. I'd have to restart the container and test to see if this works. I do have a few questions if you don't mind. Even though my app would never scale to that, but I'd like to follow the best practices as much as I can:

1. Is there an alternative way in oppose to restarting the whole container? Are you aware of anything n8n can do (script inside the container) that can just stop kafka and restart them?
2. `just disconnect the trigger from pipeline when you need to stop` - Does it mean I can just disconnect it from the nodes after and it will clear the subscription?
3. If I can't use the same topic on the same n8n instance, and I have the need to scale with multiple n8n instances in different servers, this shouldn't the problem any more right? If I understand it correctly, I can use the same topic across multiple servers? That's the point of Kafka in specific situations.
4. While we're at it, how does Kafka commit the offset? If we set it to auto commit, then what happens if the nodes after the trigger fails for some reasons? What would be the best practice here to make sure that each trigger got processed successfully at least once?

Again, thank you for your help. I'm off to testing your theory now.

**Comment by afunworm** [comment]
> @afunworm all of my suggestions, they are just like what I experienced in QUEUE MODE. (In k8s, where restarting a pod is very easy).
> 
> * If you really want to avoid restarts with 100% accuracy, i’ll suggest use some custom wrapper and call webhook of n8n on kafka subscription. eg - replace kafka trigger with webhook trigger & do not use response immediately, in place of that use respond to webhook node. In your custom wrapper we will subscribe each input topic and call n8n webhook if got success then only we will commit.
> * In above approach we can even run kafka in HA like maybe multiple custom wrapper with same consumer group id, so it,s like an perfect setup.
> * I am not sure but maybe there should be an already implemented solutions out there for calling webhook on each kafka message.
> 
> N8N can’t do like kafka reset on runtime. It’s better to handle kafka logic ourselves for better control, like i mentioned above
> 
> yea by just disconnecting i meant disconnect the next node link.
> 
> we can use same topic, but not same consumer group id. Even if topic is not same we cannot use same consumer group id.
> 
> Not sure how kafka commit offset, probably as soon as it received trigger, it commit offset automatically, like it responds to webhook in immediately option.
> 
> Extra - I dont want to make it more complex, but n8n have an very rare & hard to reproduce bug where in webhook trigger & kafka trigger
> 
> * it will repond immediately and all webhook node data becames null. So it also make whole execution data null. And when worker tries to process it, there is nothing to process for n8n workers.
> * Same happens in kafka triggers as well, bt in both cases it’s very very rare like 1 out of 1000 executions, sometimes even rare.
> * so to overcome this I’ll recommend only use webhook trigger with respond to webhook node, so if trigger node’s data is null, respond to webhook node will never run and we will never get response from n8n, it will result in timeout from caller and caller will retry for sure.
> * But here respond respond to webhook node also have some issues, like the execution order of respond to webhook node, if you use respond to webhook node in parallel ( input node of respond to webhook node have ,multiple output node ) then it behaves strange and throw error if any other node fails, means it execute in very last, which is a bad thing, so use it in a pipeline like single input and then respond to webhook node and then once we responded go for your core workflow logic.
> 
> I know things are getting complex, thats why i choose option to manually handle that rare failure if it happens.
> 
> Again all my assumptions are around n8n in QUEUE MODE
> 
> Thank You

These are all very valuable information. **Thank you so much**. I mean, in practice, it only takes a couple lines of code to write a Kafka listener that can manually commits and do things. It'd be nice to just have it reliable in n8n.

> replace kafka trigger with webhook trigger & do not use response immediately, in place of that use respond to webhook node. In your custom wrapper we will subscribe each input topic and call n8n webhook if got success then only we will commit.
Then it'd just be better off that we use HTTP trigger then? That way we handle errors & stuffs, but it also defeats the purpose of trying to use Kafka. Again, my apps will never grow that big for that to matter. It'd just be nice to be able to do this.

**Comment by Joffcom** [comment]
We’re taking some time to review and tidy up our issue tracker. It looks like this issue hasn’t had any recent activity for the past 6 months. To keep our backlog focused on current priorities, we’re closing this issue for now.

That said, this doesn’t mean the issue is forgotten! If this is still relevant or you have more information to add, feel free to reopen it or create a new issue and we’d be happy to revisit it.

Thank you for understanding and for contributing to the project!

**Comment by carmonexus** [comment]
We’re facing the same issue — people are giving up on using Confluent Kafka with n8n, which is unfortunate because it makes n8n much more robust. When can we expect a review or update of this component?
Version 106 with Bug.

I saw that when the consumer is empty it doesn't work, it seems that the connection drops and the n8n trigger doesn't reconnect.

**Comment by phmasc** [comment]
Esse erro ainda acontece! Algum planejamento?

**Comment by a3sroot** [comment]
Esse erro ainda acontece! Algum planejamento?
Version 1.114.4

**Comment by carmonexus** [comment]
I wrote a Python script to read from Kafka and call the n8n worker URL, the problem is with the reconnection:

**Summary of the issue:**
When the consumer loses its connection or partition assignment (e.g., brokers down, transport/authentication failure, or it stays alive but with no assigned partitions), the loop detects the anomaly and **shuts down the current consumer**, then **recreates it after an exponential backoff**, resuming message consumption automatically.

**Where the code handles this (automatic reconnection):**

* **Watchdogs inside the polling loop:**

  * `_should_restart_for_empty_assignment()` and `_should_restart_for_stats_empty()` inside `_poll_loop()` raise `RuntimeError(...)` to force a restart.
  * Connection health check: `c.list_topics(timeout=5.0)` inside `_poll_loop()`; any failure raises a `KafkaException` to trigger reconnection.
  * Message-level errors that require restart: in `_poll_loop()`, if `msg.error().code()` ∈ `{_ALL_BROKERS_DOWN, _AUTHENTICATION, _TRANSPORT, _FATAL}`, it raises `KafkaException(...)`.

* **Recreation cycle with exponential backoff:**

  * In `loop_listener_kafka()`: wraps `_poll_loop(c)` in a `try/except` block; when `KafkaException`, `RuntimeError`, or any `Exception` is caught, it **closes** the current consumer (`c.close()`), **waits** using exponential `backoff` (up to 60s), and then **creates a new consumer** via `get_consumer()`.

* **Graceful shutdown (prevents locks during restart/exit):**

  * `_graceful_close()` + `atexit.register(_graceful_close)` and signal handlers (`SIGTERM/SIGINT` via `_stop`) ensure the consumer (`_current_consumer.close()`) is properly closed and the producer is flushed.
[

[kafka_worker.py](https://github.com/user-attachments/files/23331075/kafka_worker.py)

](url)

**Comment by isnandar1471** [comment]
Hi, in my case, it's not listening (but not returning an error) so far. Is there a solution I can try?

**Comment by carmonexus** [comment]
I created a Web App connector in Azure that reads from Kafka and calls the WebHook; I can make it available on Git.

**Comment by VladimirBeliakov** [comment]
Looks like this issue is, to some extent, related to [this one](https://github.com/n8n-io/n8n/issues/13823). That one was fixed and closed but looks like the problem is still there.

**Comment by carmonexus** [comment]
Hi,
Look at the code for what dog inside kafka_worker.py and you'll understand how to solve the Kafka reading problem in Python and also in n8n:

https://github.com/carmomelo/n8n-install/blob/main/conector-kakfa/kafka_worker.py

# =========================
# Watchdogs and group state
# =========================
ASSIGNMENT_GRACE_SECS = 45.0 # Maximum time without assignment before recreating the consumer
STATS_GRACE_SECS = 60.0 # If stats indicate 0 partitions assigned for longer, restart
_last_assigned_ts: float = 0.0
_assigned: bool = False
_subscribe_start_ts: float = 0.0

# State coming from librdkafka statistics
_last_stats_ts: float = 0.0
_last_stats_assignment_size: int = 0
_last_stats_group_state: str = "unknown"
